### PR TITLE
Ruff: B Bugbear

### PR DESCRIPTION
**Pull request by ahms5**
See #647

### CODE REVIEW

**Review by f-brinkmann — DISMISSED** [review]
Thanks for the effort, I think adding the stacklavel to the warnings really makes thinkgs better :)

**Comment by ahms5** [comment]
merging dev into this pr required more fixes, please review again @f-brinkmann and others.

File: pyfar/dsp/dsp.py [hunk 1]
```diff
@@ -1328,14 +1328,16 @@ def find_impulse_response_delay(impulse_response, N=1):
...
+                # Use only real-valued roots
+                if np.all(np.isreal(roots)):
+                    root = roots[np.abs(roots) == np.min(np.abs(roots))]
+                    start_sample[idx] = np.squeeze(lags[argmax] + root)
+                else:
+                    start_sample[idx] = np.nan
+                    warnings.warn('Starting sample not found for channel '
+                                  f'{ch}', stacklevel=2)
+            else:
+                start_sample[idx] = np.nan
```
Comment (f-brinkmann) [inline pyfar/dsp/dsp.py:1340]: Hm, weird - did you touch the indention here and below? The browser interface shows it as changes, the VSCode interface does not...

**Review by f-brinkmann — APPROVED** [review]
Had one question for safety, but I think its fine :)

File: pyfar/io/_codec.py [hunk 2]
```diff
@@ -247,9 +247,9 @@ def _decode_object_json_aided(name, type_hint, zipfile):
     ObjType = _str_to_type(type_hint[1:])
     try:
         return ObjType._decode(obj_dict)
-    except AttributeError:
+    except AttributeError as e:
         raise NotImplementedError(
-            f'You must implement `{type}._decode` first.')
+            f'You must implement `{type}._decode` first.') from e
```
Comment (hoyer-a) [inline pyfar/io/_codec.py:252]: Why do we use `{type}` instead of something like `ObjType.__name__` here? Maybe someone can explain.

**Review by hoyer-a — COMMENTED** [review]
Looks good to me, thanks!

I just have two questions.

File: pyfar/dsp/dsp.py (hunk 1 above)
Comment (ahms5) [reply pyfar/dsp/dsp.py]: This was in one commit, which was reverted somehow (now idea but looks fine now)

File: pyfar/dsp/filter/fractional_octaves.py [hunk 3]
```diff
@@ -315,7 +316,8 @@ def _coefficients_fractional_octave_bands(
         if Wn[-1] > 1:
             warnings.warn('The upper frequency limit {} Hz is above the \
                 Nyquist frequency. Using a highpass filter instead of a \
-                bandpass'.format(np.round(freqs_upper[idx], decimals=1)))
```
[the remark this answers is not in the document]
Comment (ahms5) [reply pyfar/dsp/filter/fractional_octaves.py:318]: done

File: pyfar/io/_codec.py (hunk 2 above)
Comment (ahms5) [reply pyfar/io/_codec.py]: done, type was even wrong, thanks for noticing.

**Review by f-brinkmann — APPROVED** [review]

File: pyfar/io/_codec.py [hunk 4]
```diff
@@ -249,7 +249,7 @@ def _decode_object_json_aided(name, type_hint, zipfile):
         return ObjType._decode(obj_dict)
     except AttributeError as e:
         raise NotImplementedError(
-            f'You must implement `{type}._decode` first.') from e
+            f'You must implement `{ObjType.name}._decode` first.') from e
```
Comment (hoyer-a) [inline pyfar/io/_codec.py:252]: Should work with `ObjType.__name__`

**Review by hoyer-a — CHANGES_REQUESTED** [review]

File: pyfar/io/_codec.py [hunk 5]
```diff
@@ -247,9 +247,9 @@ def _decode_object_json_aided(name, type_hint, zipfile):
     ObjType = _str_to_type(type_hint[1:])
     try:
         return ObjType._decode(obj_dict)
-    except AttributeError:
+    except AttributeError as e:
         raise NotImplementedError(
-            f'You must implement `{type}._decode` first.')
+            f'You must implement `{ObjType.name}._decode` first.') from e
```
Comment (ahms5) [inline pyfar/io/_codec.py:252]:
```suggestion
            f'You must implement `{ObjType.__name__}._decode` first.') from e
```

File: pyfar/io/_codec.py (hunk 4 above)
Comment (ahms5) [reply pyfar/io/_codec.py]: done

**Review by hoyer-a — APPROVED** [review]
Thanks!

**Review by f-brinkmann — APPROVED** [review]

**Pull request merged**